### PR TITLE
[feat] Resume from ckpt

### PR DESCRIPTION
**Pull request by kevssim**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# PR information

实现完整训练状态的恢复——包括优化器、调度器、RNG配置以及数据集跳过

### CODE REVIEW

File: src/twinkle/model/transformers/transformers.py [hunk 1]
```diff
@@ -935,20 +977,112 @@ def load_peft_weights_for_fsdp2(model, adapter_weights, adapter_name='default'):
...
+        rng_state = torch.load(rng_path, map_location='cpu', weights_only=False)
+        random.setstate(rng_state['python_rng_state'])
+        np.random.set_state(rng_state['numpy_rng_state'])
+        torch.set_rng_state(rng_state['torch_rng_state'])
+
+        device_type = rng_state.get('device_type')
+        device_rng_state = rng_state.get('device_rng_state')
+        if device_type == 'npu' and hasattr(torch, 'npu') and torch.npu.is_available() and device_rng_state is not None:
+            torch.npu.set_rng_state(device_rng_state)
+        elif device_type == 'cuda' and torch.cuda.is_available() and device_rng_state is not None:
```
Comment (kevssim) [inline src/twinkle/model/transformers/transformers.py]: 使用platform

File: cookbook/client/twinkle/self_host/self_cognition.py [hunk 2]
```diff
@@ -99,21 +99,29 @@ def train():
     # model.set_lr_scheduler('LinearLR')
```
Comment (tastelikefeet) [inline cookbook/client/twinkle/self_host/self_cognition.py:99]: 这文件命名有一个typo

File: client_tools/client_generator.py [hunk 3]
```diff
@@ -617,6 +618,24 @@ def load(self, name: str, **kwargs) -> None:
         )
         response.raise_for_status()
 
+    def load_training_state(self, name: str, **kwargs) -> Dict[str, Any]:
```
Comment (tastelikefeet) [inline client_tools/client_generator.py]: load_training_state和read_training_progress什么区别，能否合并为一个呢

File: cookbook/client/twinkle/self_host/self_cognition.py [hunk 4]
```diff
@@ -125,13 +133,17 @@ def train():
...
                 metric = model.calculate_metric(is_training=True)
-                logger.info(f'Current is step {step} of {len(dataloader)}, metric: {metric.result}')
+                logger.info(f'Current is step {global_step} of {len(dataloader)}, metric: {metric.result}')
 
         # Step 8: Save the trained checkpoint
-        twinkle_path = model.save(name=f'twinkle-epoch-{epoch}', save_optimizer=True)
+        twinkle_path = model.save(
+            name=f'twinkle-epoch-{epoch}',
+            save_optimizer=True,
+            consumed_train_samples=consumed_train_samples,
```
Comment (tastelikefeet) [inline cookbook/client/twinkle/self_host/self_cognition.py]: dataloader.get_consumed_samples()?

Comment (tastelikefeet) [reply cookbook/client/twinkle/self_host/self_cognition.py]: 或者，dataloader.get_state()，更通用一些

Comment (tastelikefeet) [reply cookbook/client/twinkle/self_host/self_cognition.py]: 另外，这里额外测试下torchrun/ray的兼容性，还有megatron和transformers双模型的兼容性

File: docs/source_en/Components/Model/TransformersModel.md [hunk 5]
```diff
@@ -50,3 +50,16 @@ for data in dataloader:
...
   model.clip_grad_and_step(..., gradient_accumulation_steps=16)
 ```
+
+## Checkpoint and Resume
+
+`TransformersModel.save()` can save either weights only or a resumable training checkpoint.
+
+- `model.save(name, save_optimizer=True, consumed_train_samples=...)` saves weights together with optimizer, scheduler, scaler, RNG, and `trainer_state.json`.
+- `model.load(name, output_dir=..., adapter_name=...)` restores LoRA / adapter model weights.
+- `model.read_training_progress(checkpoint_dir, ...)` reads checkpoint metadata such as `cur_step`, `gradient_accumulation_steps`, and `consumed_train_samples`.
```
Comment (tastelikefeet) [inline docs/source_en/Components/Model/TransformersModel.md]: 这两个比较相似，合成一个是否合适？比如
training_progress = model.resume_from_checkpoint(xxx)
dataloader.resume_from_checkpoint(training_progress.get('dataloader'))
这样？

File: src/twinkle/model/transformers/transformers.py [hunk 6]
```diff
@@ -855,21 +860,57 @@ def save(self, name: Optional[str] = None, output_dir: Optional[str] = None, int
...
 
     def _save_optimizer(self, output_dir, **kwargs):
         adapter_name = kwargs.pop('adapter_name', _default_adapter_name)
         optimizer_config = self.optimizer_group[adapter_name]
+        optimizer = optimizer_config.optimizer
+        lr_scheduler = optimizer_config.lr_scheduler
 
+        if optimizer is not None:
+            optimizer_path = os.path.join(output_dir, 'optimizer.pt')
+            if hasattr(self.strategy, 'save_optimizer_checkpoint'):
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py]: 这里职责有点不清晰，具体为什么有的strategy有save_optimizer_checkpoint，有的又没有？
读代码的人就会感觉不理解，到底什么情况需要strategy存储

File: src/twinkle/model/transformers/transformers.py [hunk 7]
```diff
@@ -855,21 +860,57 @@ def save(self, name: Optional[str] = None, output_dir: Optional[str] = None, int
...
-            if optimizer is not None:
-                torch.save(optimizer.state_dict(), os.path.join(output_dir, 'optimizer.pt'))
             if lr_scheduler is not None:
                 torch.save(lr_scheduler.state_dict(), os.path.join(output_dir, 'scheduler.pt'))
 
+    def _save_training_state(self, output_dir, **kwargs):
+        adapter_name = kwargs.pop('adapter_name', _default_adapter_name)
+        optimizer_config = self.optimizer_group[adapter_name]
+
+        if not Platform.is_master():
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/transformers.py:900]: 这里ray和torchrun都需要确保正确，megatron部分也需要对应考虑

File: src/twinkle/model/transformers/strategy/accelerate.py [hunk 8]
```diff
@@ -124,6 +124,59 @@ def wrap_model(self, model, *args):
...
+            broadcast_from_rank0=getattr(fsdp_plugin.state_dict_config, 'rank0_only', False),
+        )
+
+    def needs_wrapped_optimizer_state(self) -> bool:
+        fsdp_plugin = self._get_fsdp_plugin()
+        return fsdp_plugin is not None and fsdp_plugin.fsdp_version == 2
+
+    def save_optimizer_checkpoint(self, model, optimizer, output_path: str):
+        fsdp_plugin = self._get_fsdp_plugin()
+        if fsdp_plugin is not None and fsdp_plugin.fsdp_version == 2:
```
Comment (tastelikefeet) [inline src/twinkle/model/transformers/strategy/accelerate.py:152]: import torch一次即可

File: cookbook/client/twinkle/self_host/self_cognition.py (hunk 2 above)
Comment (kevssim) [reply cookbook/client/twinkle/self_host/self_cognition.py:99]: 确实是 typo，但 self_congnition.py 在 main 上就已存在，是否在单独 PR 中修正更合适？

Comment (tastelikefeet) [reply cookbook/client/twinkle/self_host/self_cognition.py:99]: 可以

File: src/twinkle/model/megatron/multi_lora_megatron.py [hunk 9]
```diff
@@ -203,10 +207,71 @@ def set_lr_scheduler(self, scheduler_cls: Union[LRScheduler, Type[LRScheduler],
         self._check_adapter_valid(kwargs.get('adapter_name'))
         super().set_lr_scheduler(scheduler_cls, **kwargs)
 
+    @staticmethod
+    def _rank_local_optimizer_path(checkpoint_dir: str) -> str:
+        rank = dist.get_rank() if dist.is_initialized() else 0
+        return os.path.join(checkpoint_dir, f'optimizer_rank_{rank}.pt')
+
+    @staticmethod
+    def _get_local_training_rng_state():
```
Comment (tpx818) [inline src/twinkle/model/megatron/multi_lora_megatron.py]: 方法名统一用save和load

File: src/twinkle/model/megatron/multi_lora_megatron.py [hunk 10]
```diff
@@ -203,10 +207,71 @@ def set_lr_scheduler(self, scheduler_cls: Union[LRScheduler, Type[LRScheduler],
...
+            'np_rng_state': np.random.get_state(),
+            'torch_rng_state': torch.get_rng_state(),
+        }
+        if torch.cuda.is_available():
+            rng_state['cuda_rng_state'] = torch.cuda.get_rng_state()
+        rng_state['rng_tracker_states'] = tensor_parallel.get_cuda_rng_tracker().get_states()
+        return rng_state
+
+    @staticmethod
+    def _restore_local_training_rng_state(rng_state):
```
Comment (tpx818) [inline src/twinkle/model/megatron/multi_lora_megatron.py]: 同上

**Review by tastelikefeet — APPROVED** [review]

File: src/twinkle_client/dataloader/dataloader.py [hunk 11]
```diff
@@ -82,4 +82,44 @@ def __next__(self):
...
+        response = http_post(
+            url=f'{self.server_url}/call',
+            json_data={
+                'processor_id': self.processor_id,
+                'function': 'get_state',
+                **{},
+            }
+        )
+        response.raise_for_status()
+        return response.json()["result"]
```
Comment (Yunnglin) [inline src/twinkle_client/dataloader/dataloader.py:124]: This part of the code should be modified in the client_generator.py, and then the script should be called to generate.

File: docs/source_zh/使用指引/快速开始.md [hunk 12]
```diff
@@ -412,6 +470,7 @@ python train.py
 ```
 
 ### 远程训练
+client-server 训练场景同样支持断点续训。推荐流程是调用 `model.resume_from_checkpoint(resume_path)` 恢复权重和优化器状态，再调用 `dataloader.resume_from_checkpoint(progress['consumed_train_samples'])` 跳过已消费数据。详细示例可参考 `docs/source_zh/使用指引/服务端和客户端/Twinkle客户端.md` 和 `cookbook/client/twinkle/self_host/self_congnition.py`。
```
Comment (Yunnglin) [inline docs/source_zh/使用指引/快速开始.md]: Missing document jump link

File: docs/source_en/Usage Guide/Quick-Start.md [hunk 13]
```diff
@@ -413,6 +471,8 @@ python train.py
 
 A major feature of Twinkle is support for multi-tenant mixed training. Specifically, multiple users can use a single base model for LoRA training, which can greatly reduce server-side deployment costs.
 
+Checkpoint resumption is also supported in client-server training. The recommended flow is to call `model.resume_from_checkpoint(resume_path)` to restore weights and optimizer state, then call `dataloader.resume_from_checkpoint(progress['consumed_train_samples'])` to skip consumed data. See `docs/source_en/Usage Guide/Server and Client/Twinkle-Client.md` and `cookbook/client/twinkle/self_host/self_congnition.py`.
+
```
Comment (Yunnglin) [inline docs/source_en/Usage Guide/Quick-Start.md:475]: Missing document jump link

**Review by Yunnglin — CHANGES_REQUESTED** [review]

File: src/twinkle_client/dataloader/dataloader.py (hunk 11 above)
Comment (kevssim) [reply src/twinkle_client/dataloader/dataloader.py:124]: The DataLoader client file was regenerated, not manually edited.

These methods are generated from the new `@remote_function()` methods in `src/twinkle/dataloader/dataloader.py`; no generator logic change is needed for this case because `client_tools/client_generator.py` already scans remote functions and emits the client wrappers.

File: src/twinkle/model/transformers/transformers.py [hunk 14]
```diff
@@ -927,10 +963,11 @@ def load_peft_weights_for_fsdp2(model, adapter_weights, adapter_name='default'):
...
-                    if key in model_sd:
-                        param = model_sd[key]
+                    model_key = key
+                    if f'.{adapter_name}.weight' not in model_key:
+                        model_key = model_key.replace('.weight', f'.{adapter_name}.weight')
+                    if model_key in model_sd:
+                        param = model_sd[model_key]
                         if isinstance(param, DTensor) and not isinstance(value, DTensor):
                             value = distribute_tensor(value.to(param.device), param.device_mesh, param.placements)
                     converted_weights[key] = value
```
[the remark this answers is not in the document]
Comment (kevssim) [reply src/twinkle/model/transformers/transformers.py:973]: The current code is correct. set_peft_model_state_dict internally inserts the adapter name into keys. model_key is only used to look up DTensor metadata from model.state_dict(). Storing under model_key would double-insert the adapter name and break loading.

File: src/twinkle/dataloader/dataloader.py [hunk 15]
```diff
@@ -114,13 +135,57 @@ def __iter__(self):
...
-                self.dataloader.batch_sampler.sampler, self.dataset, max_retries=self.max_retries)
-            self.dataloader.batch_sampler = DeviceMeshSampler(self.dataloader.batch_sampler, self.device_mesh,
-                                                              self.min_batch_size)
-        elif self.dataloader.sampler is not None:
-            self.dataloader.sampler = RetrySampler(self.dataloader.sampler, self.dataset, max_retries=self.max_retries)
+        return self._tracking_iter(_iter)
+
+    def _tracking_iter(self, inner):
+        for batch in inner:
+            self._consumed_train_samples += self.batch_size
```
[the remark this answers is not in the document]
Comment (kevssim) [reply src/twinkle/dataloader/dataloader.py:142]: _consumed_train_samples counts samples for checkpoint resume via RetrySampler(skip_samples=N). The last batch may over-count by up to batch_size - 1 when drop_last=False, but this only matters if resuming at the exact end of an epoch — and RetrySampler clamps to max(len(dataset) - skip_samples, 0), so it won't error. Fixing this is non-trivial since batch structure varies (dict, tuple, tensor) and len(batch) doesn't reliably give sample count. Low priority, not worth the complexity.

**Review by Yunnglin — APPROVED** [review]
